### PR TITLE
Fix bugs in New Parquet Reader

### PR DESCRIPTION
**Pull request by zhenxiao**
Fix 3 things in New Parquet Reader:
1. ParquetReader reading incorrect row group metadata
2. LazyRead does not do necessary skips
3. ParquetReader and ParquetFileReader could be merged into one class

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetFileReader.java [hunk 1]
```diff
@@ -80,7 +68,9 @@ public ParquetColumnChunkPageReader readColumn(ColumnDescriptor columnDescriptor
     {
         checkArgument(currentBlockMetadata.getRowCount() > 0, "Row group having 0 rows");
 
-        ColumnChunkMetaData metadata = columnMetadata.get(columnDescriptor);
+        ColumnChunkMetaData metadata = getColumnChunkMetaData(columnDescriptor);
+        checkArgument(metadata != null, "Could not find column metadata in the parquet file");
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetFileReader.java]: why not put this in `getColumnChunkMetaData` and then the method would never return null

Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetFileReader.java]: Should this be some kind of Presto corruption exception (like we have for ORC)?

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 2]
```diff
@@ -45,7 +46,8 @@
     private long currentPosition;
     private long currentGroupRowCount;
     private long nextRowInGroup;
-    private Map<ColumnDescriptor, ParquetColumnReader> columnReadersMap = new HashMap<>();
+    private long currentRowGroup;
+    private Map<ColumnIndex, ParquetColumnReader> columnReadersMap = new HashMap<>();
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: For a big file will this map become huge?  If so, maybe clear the map in advanceToNextRowGroup

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 3]
```diff
@@ -126,13 +129,58 @@ public Block readBlock(ColumnDescriptor columnDescriptor, int vectorSize, Type t
...
+        private final long rowGroupNumber;
+        private final ColumnDescriptor columnDescriptor;
+
+        public ColumnIndex(long rowGroupNumber, ColumnDescriptor columnDescriptor)
+        {
+            this.rowGroupNumber = rowGroupNumber;
+            this.columnDescriptor = columnDescriptor;
+        }
+
+        public long getRowGroupNumber()
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: If these aren't called, just remove them for now.  We can always add them back later

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java [hunk 4]
```diff
@@ -126,13 +129,58 @@ public Block readBlock(ColumnDescriptor columnDescriptor, int vectorSize, Type t
...
+            if (this == obj) {
+                return true;
+            }
+
+            if (obj == null || !getClass().equals(obj.getClass())) {
+                return false;
+            }
+
+            ColumnIndex other = (ColumnIndex) obj;
+            return rowGroupNumber == other.getRowGroupNumber() &&
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java]: use the direct fields of `other`

**Comment by zhenxiao** [comment]
thank you @dain , get comments addressed
Just found no need to make ParquetColumnReader indexed by both rowGroupNumber and ColumnDescriptor, just indexed by ColumnDescriptor is OK

**Comment by zhenxiao** [comment]
@dain I updated this PR with more fixes in the new Parquet Reader, your comments and suggestions are appreciated

**Comment by zhenxiao** [comment]
also merge the testcase PR here

**Comment by zhenxiao** [comment]
Add another bug fix, following @electrum 's handling of missing columns where columns were added to the metastore after table data was written:
https://github.com/facebook/presto/commit/e4b6ee534120c39d698f1ac171c74b30c9a4816f

Also fix the handling of missing columns in the new Parquet Reader

**Comment by dain** [comment]
I'm repeating this comment from one of the commits....

For all of the Parquet code, can you take a look at the calls like `checkArgument` and if they are actually checks for file corruption (as opposed to programming errors), switch them to `ParquetCorruptionException`.

**Comment by dain** [comment]
Additionally, look at all uses of the new `ParquetCorruptionException` and remove unnecessary call to `toString` on the args

**Comment by dain** [comment]
This PR seems to have `this.` on most uses of class fields.  Generally, we don't add an unnecessary `this.` to field accesses unless it helps with readability (you see this in some constructors).

**Comment by dain** [comment]
Can you scan through all the commits and fix any place where the method arguments are aligned, instead of using two indents (8 spaces)?

**Comment by dain** [comment]
For the new `AbstractTestParquetReader` code, the goal should be to test data streams that cover all of the various encodings in Parquet.  My guess is you don't need cases like `testLongStrideDictionary` and `testLongPatchedBase` since these are specifically designed for funky ORC encodings.  Additionally, it is possible you are missing some to stress your code.  I'd used the code coverage tool in IntelliJ to verify I had cases for all the ORC encodings.

**Comment by dain** [comment]
My comments are mostly style/formatting.  Let me know when this is updated and I'll get it in.

**Comment by zhenxiao** [comment]
Thank u so much @dain 
I get comments addressed

**Comment by dain** [comment]
I will land this after the next release goes out (they are working on it now).

**Comment by dain** [comment]
Merged, thanks!

**Pull request closed without merging**